src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+if the value doesn't have the same type.
+
+The getter name can be omitted and inferred from the [variable pattern](#variable)
+or [identifier pattern](#identifier) in the field subpattern:
+
+```dart
+var Point(:var x, :var y) = Point(1, 2);
+```
+
+Object patterns don't require the pattern to match the entire object.
+If an object has extra fields that the pattern doesn't destructure, it can still match.
+
+## Parenthesized
+
+`(subpattern)`
+
+Like parenthesized expressions, parentheses in a pattern let you control pattern
+precedence and insert a lower precedence pattern where a higher precedence one is
+expected.
+
+## Record
+
+`(subpattern1, subpattern2)`
+
+`(x: subpattern1, y: subpattern2)`
+
+Record patterns match a [record][] object and destructure its fields.
+If the value isn't a record with the same [shape][] as the pattern, the match
+fails. Otherwise, the field subpatterns are matched against the corresponding
+fields in the record.
+
+```dart
+// Variable:
+var (untyped: untyped, typed: int typed) = // ...
+var (:untyped, :int typed) = // ...
+
+switch (obj) {
+  case (untyped: var untyped, typed: int typed): // ...
+  case (:var untyped, :int typed): // ...
+}
+
+// Null-check and null-assert:
+switch (obj) {
+  case (checked: var checked?, asserted: var asserted!): // ...
+  case (:var checked?, :var asserted!): // ...
+}
+
+// Cast:
+var (field: field as int) = // ...
+var (:field as int) = // ...
+```
+
+The getter name can be omitted and inferred from the [variable pattern](#variable)
+or [identifier pattern](#identifier) in the field subpattern.
+
+Record patterns require that the pattern match the entire record.
+
+## Relational
+
+`== expression`
+
+`< expression`
+
+Relational patterns compare the matched value to a given constant using any of
+the equality or relational operators: `==`, `!=`, `<`, `>`, `<=`, and `>=`.
+
+The pattern matches when calling the appropriate operator on the matched value
+with the constant as an argument returns `true`.
+
+Relational patterns are useful for matching on numeric ranges, especially when
+combined with the [logical-and pattern](#logical-and):
+
+```dart
+String asciiCharType(int char) {
+  const space = 32;
+  const zero = 48;
+  const nine = 57;
+
+  return switch (char) {
+    < space => 'control',
+    == space => 'space',
+    > space && < zero => 'punctuation',
+    >= zero && <= nine => 'digit'
+    // Etc...
+  }
+}
+```
+
+## Variable
+
+`var bar, String str, final int _`
+
+Variable patterns bind new variables to values that have been matched or destructured. 
+They usually occur as part of a [destructuring pattern][destructure] in order to
+capture a destructured value.
+
+The variables are in scope in a region of code that is only reachable when the
+pattern has matched.
+
+```dart
+switch ((1, 2)) {
+  case (var a, var b): // ...
+  // var a and var b are variable patterns that bind to 1 and 2, respectively.
+  // They are in scope in the case body.
+}
+```
```

Comment:
Maybe mention that a typed variable pattern only matches if the matched value has the declared type (and otherwise the match fails)?

Reply:
Nice, I actually had that as part of the definition originally but took it out because I wasn't sure if it was useful or not. Thanks for confirming!